netroute_bsd.go

```diff
@@ -13,137 +13,378 @@
+			return iface.Index, nil
 		}
-		if src, err := toIPAddr(m.Addrs[5]); err == nil {
-			ipn = &net.IPNet{IP: src, Mask: net.CIDRMask(8*len(src), 8*len(src))}
-			routeInfo.Src = ipn
-			routeInfo.PrefSrc = src
-			if m.Flags&0x2 != 0 /* RTF_GATEWAY */ {
-				routeInfo.Src.Mask = net.CIDRMask(0, 8*len(routeInfo.Src.IP))
-			}
+	}
+	return -1, fmt.Errorf("no interface found for MAC: %s", hwAddr.String())
+}
+
+// getIfIndex determines the interface index based on the provided MAC address and/or IP address.
+// If both are provided, it ensures they resolve to the same interface.
+// Returns -1 if neither MAC nor IP is provided, or an error if they resolve to different interfaces.
+func getIfIndex(MACAddr net.HardwareAddr, ip net.IP) (int, error) {
+	ipIndex := -1
+	macIndex := -1
+	var err error
+	if ip == nil && MACAddr == nil {
+		return -1, nil
+	}
+
+	if ip != nil {
+		ipIndex, err = ipToIfIndex(ip)
+		if err != nil {
+			return -1, fmt.Errorf("failed to find interface with IP: %s", ip.String())
 		}
-		routeInfo.OutputIface = uint32(m.Index)
+	}
 
-		switch m.Addrs[0].(type) {
-		case *route.Inet4Addr:
-			rtr.v4 = append(rtr.v4, routeInfo)
-		case *route.Inet6Addr:
-			rtr.v6 = append(rtr.v6, routeInfo)
+	if MACAddr != nil {
+		macIndex, err = macToIfIndex(MACAddr)
+		if err != nil {
+			return -1, fmt.Errorf("failed to find interface with MAC: %s", MACAddr.String())
 		}
 	}
-	sort.Sort(rtr.v4)
-	sort.Sort(rtr.v6)
-	ifaces, err := net.Interfaces()
+
+	switch {
+	case (ipIndex >= 0 && macIndex >= 0) && (macIndex != ipIndex):
+		return -1, fmt.Errorf("given MAC address and source IP resolve to different interfaces")
+	case (ipIndex >= 0 && macIndex >= 0) && (macIndex == ipIndex):
+		return ipIndex, nil
+	case ipIndex >= 0:
+		return ipIndex, nil
+	case macIndex >= 0:
+		return macIndex, nil
+	default:
+		return -1, fmt.Errorf("no index found for given ip and/or MAC address")
+	}
+}
+
+// composeRouteMsg creates a RTM_GET RouteMessage for querying the routing table.
+// It takes the process ID, optional MAC address, optional source IP, and destination IP.
+// The function determines the appropriate interface index if MAC or source IP is provided.
+func composeRouteMsg(id uintptr, seq int, MACAddr net.HardwareAddr, src, dst net.IP) (*route.RouteMessage, error) {
+	dstAddr := toRouteAddr(dst)
+	if dstAddr == nil {
+		return nil, fmt.Errorf("failed to parse dst: %#v", dst)
+	}
+
+	msg := &route.RouteMessage{
+		Version: syscall.RTM_VERSION,
+		Type:    unix.RTM_GET,
+		ID:      id,
+		Seq:     seq,
+		Addrs: []route.Addr{
+			dstAddr,
+			nil,
+			nil,
+			nil,
+			&route.LinkAddr{},
+		},
+	}
+
+	ifIndex, err := getIfIndex(MACAddr, src)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine ifIndex: %s", err)
+	}
+
+	if ifIndex >= 0 {
+		msg.Flags = RTF_IFSCOPE
+		msg.Index = ifIndex
+	}
+
+	return msg, nil
+}
+
+func (r *bsdRouter) getSeq() int {
+	r.seq.Add(1)
+	return int(r.seq.Load())
```

Comment:
The gap between the add and the load can be squashed.

```suggestion
	return int(r.seq.Add(1))
```